pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java

```diff
@@ -133,50 +168,71 @@ public PinotTaskManager getTaskManager() {
-    } catch (Exception e) {
-      Utils.rethrowException(e);
+    switch (_controllerMode) {
+      case DUAL:
+        setUpHelixController();
+        setUpPinotController();
+        break;
+      case PINOT_ONLY:
+        setUpPinotController();
+        break;
+      case HELIX_ONLY:
+        setUpHelixController();
+        break;
+      default:
+        LOGGER.error("Invalid mode: " + _controllerMode);
     }
 
-    LOGGER.info("Initializing SegmentFetcherFactory");
-    try {
-      SegmentFetcherFactory.getInstance().init(segmentFetcherFactoryConfig);
-    } catch (Exception e) {
-      throw new RuntimeException("Caught exception while initializing SegmentFetcherFactory", e);
-    }
+    _controllerMetrics.initializeGlobalMeters();
+  }
 
-    LOGGER.info("Initializing PinotCrypterFactory");
-    try {
-      PinotCrypterFactory.init(pinotCrypterConfig);
-    } catch (Exception e) {
-      throw new RuntimeException("Caught exception while initializing PinotCrypterFactory", e);
+  private void setUpHelixController() {
+    // Register and connect instance as Helix controller.
+    _helixControllerManager = HelixSetupUtils
+        .setup(_helixClusterName, _helixZkURL, _instanceId, _isUpdateStateModel, _enableBatchMessageMode);
+
+    // Emit helix controller metrics
+    _controllerMetrics.addCallbackGauge("helix.connected", () -> _helixControllerManager.isConnected() ? 1L : 0L);
+    _controllerMetrics.addCallbackGauge("helix.leader", () -> _helixControllerManager.isLeader() ? 1L : 0L);
+    _helixControllerManager.addPreConnectCallback(
+        () -> _controllerMetrics.addMeteredGlobalValue(ControllerMeter.HELIX_ZOOKEEPER_RECONNECTS, 1L));
+  }
+
+  private void setUpPinotController() {
+    // Note: Right now we don't allow pinot-only mode to be used in production yet.
+    // Now we only have this mode used in tests.
+    // TODO: Remove this logic once all the helix separation PRs are committed.
+    if (!isPinotOnlyModeSupported()) {
+      throw new RuntimeException("Pinot only controller currently isn't supported in production yet.");
     }
 
+    // Start all components
+    initPinotFSFactory();
+    initSegmentFetcherFactory();
+    initPinotCrypterFactory();
+
     LOGGER.info("Starting Pinot Helix resource manager and connecting to Zookeeper");
-    _helixResourceManager.start();
-    final HelixManager helixManager = _helixResourceManager.getHelixZkManager();
+    _helixParticipantManager = registerAndConnectAsHelixParticipant();
+    _helixResourceManager.start(_helixParticipantManager);
 
     LOGGER.info("Init controller leadership manager");
-    ControllerLeadershipManager.init(helixManager);
+    // Note: Currently leadership depends on helix controller, thus assign helixControllerManager to ControllerLeadershipManager.
+    // In the future when Helix separation is completed, leadership only depends on the master in leadControllerResource.
+    if (_helixControllerManager != null) {
```

Comment:
Move this to `setUpHelixController()`? Also add a TODO for this

Reply:
Right now Pinot controller (pinot only and dual modes) still need this logic for realtime segment consumption. I think for now it's best to leave it here. Will add TODO in the comment.

Reply:
For pinot only mode, `_helixControllerManager` is null right?

Reply:
Yes, and for Pinot only mode, we won't touch this line since it's gotten an exception from `isPinotOnlyModeSupported()` and exit the app.

Reply:
And that's why we should move this into method `setUpHelixController()` right? The ControllerLeadershipManager is for Helix controller, not Pinot controller.

Reply:
Added a `TODO` in the comment. `ControllerLeadershipManager` will be removed once everything is ready.